source/open_node/ast.py

```diff
@@ -0,0 +1,1017 @@
+from __future__ import annotations
+
+from dataclasses import dataclass, field
+from typing import List, Optional, Union
+from enum import Enum, auto
+
+
+@dataclass
+class Span:
+    """
+    Start and end line/column in the source file
+    We use the Antlr convention. The starting line number is 1 and starting column number is 0.
+    """
+
+    start_line: int
+    start_column: int
+    end_line: int
+    end_column: int
+
+
+@dataclass
+class OpenNode:
+    """Base class for all OpenQASM 3 nodes"""
+
+    span: Optional[Span] = field(init=False, default=None, compare=False)
+    """
+    The span(location) of the node in the source code.
+    Because not all the nodes are generated from source, the span is optional.
+    To make it easier to write unit test, we exclude span from the generated __eq__().
+    """
+
+
+@dataclass
+class Program(OpenNode):
+    """
+    An entire OpenQASM 3 program represented by a list of top level statements
+    """
+
+    statements: List[Statement]
+    version: dict[str, int] = field(init=False, default_factory=dict, compare=False)
+    io_variables: List[IODeclaration] = field(init=False, default_factory=list)
```

Comment:
Are the IODeclarations also part of the statements list above?

Reply:
In the current ANTLR grammar, IODeclarations are part of the header, not a statement.